addons/mrp/models/mrp_bom.py

```diff
@@ -261,15 +270,20 @@ def check_cycle(v, visited, recStack, graph):
                 continue
 
             line_quantity = current_qty * current_line.product_qty
-            bom = self._bom_find(product=current_line.product_id, picking_type=picking_type or self.picking_type_id, company_id=self.company_id.id, bom_type='phantom')
+            if not current_line.product_id in product_boms:
+                product_boms.update(self._get_product2bom(products_to_search, bom_type='phantom', picking_type=picking_type or self.picking_type_id))
+                products_to_search = self.env['product.product']
+            bom = product_boms.get(current_line.product_id)
             if bom:
                 converted_line_quantity = current_line.product_uom_id._compute_quantity(line_quantity / bom.product_qty, bom.product_uom_id)
-                bom_lines = [(line, current_line.product_id, converted_line_quantity, current_line) for line in bom.bom_line_ids] + bom_lines
+                bom_lines = bom_lines + [(line, current_line.product_id, converted_line_quantity, current_line) for line in bom.bom_line_ids]
```

Comment:
Why this change ? (if it is legit, maybe use `+=`)

Reply:
This allows to group the lines product_id of the same hierarchy level together, reducing the number of calls to _get_product2bom. If we have 1 bom with 4 bom_lines, each one with a bom with 2 bom_lines, earch one with a bom, kinda like this:
<pre>
.bom 1
+--- bom 2
|      +--- bom 6
|      +--- bom 7
+--- bom 3
|      +--- bom 8
|      +--- bom 9
+--- bom 4
|      +--- bom 10
|      +--- bom 11
+--- bom 5
|      +--- bom 12
|      +--- bom 13
</pre>

The last level will be grouped together, reducing the number of calls. In this example, the number of calls is theoretically equal to 3 (the height of the tree+1) while it would be 13 without the change (1 call for the first level after root and 1 for each sibling group in the tree). I say theoretically because with the backport of 14.3 there are more calls to product2boms but most of them enters the first if condition.

In practice I don't think that it will change the perf that much because most of the time there are only one or two levels in the bom hierarchy but it is still worth it I think.

PS: I changed the line to `bom_lines += ...`